File: 05/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Point struct {
	x, y, z int
	idx     int
}

type Edge struct {
	from, to int
	weight   int
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 1<<20)
	writer := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer writer.Flush()

	// Читаем N
	line, _ := reader.ReadString('\n')
	N, _ := strconv.Atoi(strings.TrimSpace(line))

	// Читаем точки
	points := make([]Point, N)
	for i := 0; i < N; i++ {
		line, _ = reader.ReadString('\n')
		parts := strings.Fields(strings.TrimSpace(line))
		x, _ := strconv.Atoi(parts[0])
		y, _ := strconv.Atoi(parts[1])
		z, _ := strconv.Atoi(parts[2])
		points[i] = Point{x: x, y: y, z: z, idx: i}
	}

	result := solveMST(points)
	writer.WriteString(fmt.Sprintf("%d\n", result))
}

func find(parent []int, x int) int {
	if parent[x] != x {
		parent[x] = find(parent, parent[x])
	}
	return parent[x]
}

func union(parent, rank []int, x, y int) {
	if rank[x] < rank[y] {
		parent[x] = y
	} else if rank[x] > rank[y] {
		parent[y] = x
	} else {
		parent[y] = x
		rank[x]++
	}
}

// solveMST находит минимальное остовное дерево для заданных точек
func solveMST(points []Point) int64 {
	N := len(points)

	// Строим рёбра: для каждой координаты сортируем точки и добавляем рёбра между соседями
	edges := make([]Edge, 0, 3*N) // Максимум 3*(N-1) рёбер

	// Сортируем по x и добавляем рёбра между соседями
	sortedByX := make([]Point, N)
	copy(sortedByX, points)
	sort.Slice(sortedByX, func(i, j int) bool {
		return sortedByX[i].x < sortedByX[j].x
	})
	for i := 0; i < N-1; i++ {
		cost := min(
			abs(sortedByX[i].x-sortedByX[i+1].x),
			min(
				abs(sortedByX[i].y-sortedByX[i+1].y),
				abs(sortedByX[i].z-sortedByX[i+1].z),
			),
		)
		edges = append(edges, Edge{
			from:   sortedByX[i].idx,
			to:     sortedByX[i+1].idx,
			weight: cost,
		})
	}

	// Сортируем по y и добавляем рёбра между соседями
	sortedByY := make([]Point, N)
	copy(sortedByY, points)
	sort.Slice(sortedByY, func(i, j int) bool {
		return sortedByY[i].y < sortedByY[j].y
	})
	for i := 0; i < N-1; i++ {
		cost := min(
			abs(sortedByY[i].x-sortedByY[i+1].x),
			min(
				abs(sortedByY[i].y-sortedByY[i+1].y),
				abs(sortedByY[i].z-sortedByY[i+1].z),
			),
		)
		edges = append(edges, Edge{
			from:   sortedByY[i].idx,
			to:     sortedByY[i+1].idx,
			weight: cost,
		})
	}

	// Сортируем по z и добавляем рёбра между соседями
	sortedByZ := make([]Point, N)
	copy(sortedByZ, points)
	sort.Slice(sortedByZ, func(i, j int) bool {
		return sortedByZ[i].z < sortedByZ[j].z
	})
	for i := 0; i < N-1; i++ {
		cost := min(
			abs(sortedByZ[i].x-sortedByZ[i+1].x),
			min(
				abs(sortedByZ[i].y-sortedByZ[i+1].y),
				abs(sortedByZ[i].z-sortedByZ[i+1].z),
			),
		)
		edges = append(edges, Edge{
			from:   sortedByZ[i].idx,
			to:     sortedByZ[i+1].idx,
			weight: cost,
		})
	}

	// Сортируем рёбра по весу
	sort.Slice(edges, func(i, j int) bool {
		return edges[i].weight < edges[j].weight
	})

	// Алгоритм Крускала с DSU
	parent := make([]int, N)
	rank := make([]int, N)
	for i := 0; i < N; i++ {
		parent[i] = i
		rank[i] = 0
	}

	var totalCost int64
	edgesUsed := 0

	for _, edge := range edges {
		if edgesUsed == N-1 {
			break
		}
		fromRoot := find(parent, edge.from)
		toRoot := find(parent, edge.to)
		if fromRoot != toRoot {
			union(parent, rank, fromRoot, toRoot)
			totalCost += int64(edge.weight)
			edgesUsed++
		}
	}

	return totalCost
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

File: 05/main.rs
use std::io::{self, BufRead, BufWriter, Write};

#[derive(Clone, Copy)]
struct Point {
    x: i32,
    y: i32,
    z: i32,
    idx: usize,
}

struct Edge {
    from: usize,
    to: usize,
    weight: i32,
}

fn main() {
    let stdin = io::stdin();
    let mut reader = stdin.lock();
    let stdout = io::stdout();
    let mut writer = BufWriter::new(stdout.lock());

    // Читаем N
    let mut line = String::new();
    reader.read_line(&mut line).unwrap();
    let n: usize = line.trim().parse().unwrap();

    // Читаем точки
    let mut points = Vec::with_capacity(n);
    for i in 0..n {
        line.clear();
        reader.read_line(&mut line).unwrap();
        let parts: Vec<&str> = line.trim().split_whitespace().collect();
        let x: i32 = parts[0].parse().unwrap();
        let y: i32 = parts[1].parse().unwrap();
        let z: i32 = parts[2].parse().unwrap();
        points.push(Point { x, y, z, idx: i });
    }

    let result = solve_mst(&mut points);
    writeln!(writer, "{}", result).unwrap();
    writer.flush().unwrap();
}

fn find(parent: &mut [usize], x: usize) -> usize {
    if parent[x] != x {
        parent[x] = find(parent, parent[x]);
    }
    parent[x]
}

fn union(parent: &mut [usize], rank: &mut [usize], x: usize, y: usize) {
    if rank[x] < rank[y] {
        parent[x] = y;
    } else if rank[x] > rank[y] {
        parent[y] = x;
    } else {
        parent[y] = x;
        rank[x] += 1;
    }
}

// solve_mst находит минимальное остовное дерево для заданных точек
fn solve_mst(points: &mut [Point]) -> i64 {
    let n = points.len();

    // Строим рёбра: для каждой координаты сортируем точки и добавляем рёбра между соседями
    let mut edges = Vec::with_capacity(3 * n);

    // Сортируем по x и добавляем рёбра между соседями
    let mut sorted_by_x = points.to_vec();
    sorted_by_x.sort_by_key(|p| p.x);
    for i in 0..n - 1 {
        let cost = (sorted_by_x[i].x - sorted_by_x[i + 1].x)
            .abs()
            .min((sorted_by_x[i].y - sorted_by_x[i + 1].y).abs())
            .min((sorted_by_x[i].z - sorted_by_x[i + 1].z).abs());
        edges.push(Edge {
            from: sorted_by_x[i].idx,
            to: sorted_by_x[i + 1].idx,
            weight: cost,
        });
    }

    // Сортируем по y и добавляем рёбра между соседями
    let mut sorted_by_y = points.to_vec();
    sorted_by_y.sort_by_key(|p| p.y);
    for i in 0..n - 1 {
        let cost = (sorted_by_y[i].x - sorted_by_y[i + 1].x)
            .abs()
            .min((sorted_by_y[i].y - sorted_by_y[i + 1].y).abs())
            .min((sorted_by_y[i].z - sorted_by_y[i + 1].z).abs());
        edges.push(Edge {
            from: sorted_by_y[i].idx,
            to: sorted_by_y[i + 1].idx,
            weight: cost,
        });
    }

    // Сортируем по z и добавляем рёбра между соседями
    let mut sorted_by_z = points.to_vec();
    sorted_by_z.sort_by_key(|p| p.z);
    for i in 0..n - 1 {
        let cost = (sorted_by_z[i].x - sorted_by_z[i + 1].x)
            .abs()
            .min((sorted_by_z[i].y - sorted_by_z[i + 1].y).abs())
            .min((sorted_by_z[i].z - sorted_by_z[i + 1].z).abs());
        edges.push(Edge {
            from: sorted_by_z[i].idx,
            to: sorted_by_z[i + 1].idx,
            weight: cost,
        });
    }

    // Сортируем рёбра по весу
    edges.sort_by_key(|e| e.weight);

    // Алгоритм Крускала с DSU
    let mut parent: Vec<usize> = (0..n).collect();
    let mut rank = vec![0; n];

    let mut total_cost = 0i64;
    let mut edges_used = 0;

    for edge in edges {
        if edges_used == n - 1 {
            break;
        }
        let from_root = find(&mut parent, edge.from);
        let to_root = find(&mut parent, edge.to);
        if from_root != to_root {
            union(&mut parent, &mut rank, from_root, to_root);
            total_cost += edge.weight as i64;
            edges_used += 1;
        }
    }

    total_cost
}





